docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,448 @@
+- Deprecate or remove the ClusterResourceSet experimental feature.
+
+## Proposal
+
+This document introduces the concept of a ClusterAddonProvider, a new pluggable component to be deployed on the Management Cluster. A ClusterAddonProvider must be implemented using a specified package management tool and will act as an intermediary/proxy between Cluster API and the chosen package management solution.
+
+This proposal also introduces a ClusterAddonProvider for [Helm](https://helm.sh/) as a concrete example of how a ClusterAddonProvider would work with a package management tool. Users can configure ClusterAddonProvider for Helm through a new Kubernetes CRD called HelmChartProxy to be deployed on the Management Cluster. HelmChartProxy will act as an intermediary (proxy) between Cluster API and the Helm to install packages onto workload clusters. The implementation of ClusterAddonProvider for Helm can be used as a reference to implement ClusterAddonProviders for other package managers, i.e. ClusterAddonProvider for kapp.
+
+### User Stories
+
+#### Story 1
+
+As a user/operator, I would like Cluster add-on providers to automatically include or install the package management tool.
+
+#### Story 2
+
+As a user/operator, I would like Cluster add-on providers to expose configurations native to the package management tool to install Cluster add-ons into the newly created Workload Cluster.
+
+#### Story 3
+
+As a user/operator, I would like Cluster add-on providers to leverage existing capabilities of the package management tool, such as identifying the proper version of Cluster add-ons package to install onto the workload Cluster.
+
+#### Story 4
+
+As a user/operator, I would like Cluster add-on providers to use information available in the Management Cluster (e.g. service or pod CIDR on the Cluster object) for customizing add-ons configurations specifically for the newly created workload Cluster.
+
+#### Story 5
+
+As a user/operator, I would like Cluster add-on providers to update add-ons configurations when information available in the Management Cluster (e.g add-on configuration) changes.
+
+### ClusterAddonProvider Functionality
+
+A ClusterAddonProvider is a new component for bringing add-ons into the Cluster API lifecycle. It serves as a general solution to Cluster API Add-ons with multiple concrete implementations, similar to the relationship between Cluster API and infrastructure providers. It aims to do so handle add-on lifecycle management in a holistic way by providing the following capabilities:
+
+- Provide or install a package management tool
+- Determine which add-ons to install on each Cluster
+- Provide a default configuration for each add-on
+- Configure an add-on by resolving or inherit information from the Cluster configuration (i.e. service or pod CIDR for CNI)
+- Install add-ons during cluster creation (immediately after the API server is installed). This operation usually brings in some additional requirements such as: 
```

Comment:
>- Install add-ons during cluster creation (immediately after the API server is installed)

how is this ensured?

Reply:
It's not currently ensured in this first iteration. Maybe @fabriziopandini can comment on what he had in mind?

Reply:
I think that if we phrase this as (immediately after the API server is available) this will be more clear.
WRT to implementation, we can look at control-plane or cluster conditions for now; in a future iteration we will rely on lifecycle hooks